ProjectLighthouse/Helpers/SlotHelper.cs

```diff
@@ -0,0 +1,102 @@
+﻿#nullable enable
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using LBPUnion.ProjectLighthouse.Levels;
+using LBPUnion.ProjectLighthouse.PlayerData.Profiles;
+using Microsoft.EntityFrameworkCore;
+
+namespace LBPUnion.ProjectLighthouse.Helpers;
+
+public static class SlotHelper
+{
+
+    public static SlotType ParseType(string? slotType)
+    {
+        if (slotType == null) return SlotType.Unknown;
+        return slotType switch
+        {
+            "developer" => SlotType.Developer,
+            "user" => SlotType.User,
+            "moon" => SlotType.Moon,
+            "pod" => SlotType.Pod,
+            "local" => SlotType.Local,
+            _ => SlotType.Unknown,
+        };
+
+    }
+
+    public static bool IsTypeInvalid(string? slotType)
+    {
+        if (slotType == null) return true;
+        return slotType switch
+        {
+            "developer" => false,
+            "user" => false,
+            _ => true,
+        };
+    }
+
+    private static readonly SemaphoreSlim semaphore = new(1, 1); 
+
+    public static async Task<int> GetPlaceholderSlotId(Database database, int guid, SlotType slotType)
+    {
+        int slotId = await database.Slots.Where(s => s.Type == slotType && s.InternalSlotId == guid).Select(s => s.SlotId).FirstOrDefaultAsync();
+        if (slotId != 0) return slotId;
+
+        await semaphore.WaitAsync();
+        try
+        {
+            // if two requests come in at the same time for the same story level which hasn't been generated
+            // one will wait for the lock to be released and the second will be caught by this second check
+            slotId = await database.Slots.Where(s => s.Type == slotType && s.InternalSlotId == guid).Select(s => s.SlotId).FirstOrDefaultAsync();
```

Comment:
For huge queries like this it's best to split it up into multiple lines, like so:

```suggestion
            slotId = await database.Slots
                           .Where(s => s.Type == slotType && s.InternalSlotId == guid)
                           .Select(s => s.SlotId)
                           .FirstOrDefaultAsync();
```